src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala

```diff
@@ -333,13 +342,60 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
             case _ => assignableTo.errors(l, UNTYPED_INT_CONST)(n) ++ assignableTo.errors(r, UNTYPED_INT_CONST)(n)
           }
           case (_: PAdd, StringT, StringT) => noMessages
-          case (_: PAdd | _: PSub | _: PMul | _: PMod | _: PDiv, l, r) =>
-            if (l == PermissionT || r == PermissionT || getTypeFromCtxt(n.asInstanceOf[PNumExpression]).contains(PermissionT)) {
+          case (_: PAdd | _: PSub | _: PMul | _: PMod | _: PDiv | _: PBitAnd | _: PBitOr | _: PBitXor | _: PBitClear, l, r) =>
+            if (l == PermissionT || r == PermissionT || getTypeFromCtxt(n).contains(PermissionT)) {
               assignableTo.errors(l, PermissionT)(n) ++ assignableTo.errors(r, PermissionT)(n)
             } else {
-              assignableTo.errors(l, UNTYPED_INT_CONST)(n) ++ assignableTo.errors(r, UNTYPED_INT_CONST)(n) ++
-                numExprWithinTypeBounds(n.asInstanceOf[PNumExpression])
+              val lIsInteger = assignableTo.errors(l, UNTYPED_INT_CONST)(n)
+              val rIsInteger = assignableTo.errors(r, UNTYPED_INT_CONST)(n)
+              val typesAreMergeable = mergeableTypes.errors(l, r)(n)
+              val exprWithinBounds = {
+                if(typesAreMergeable.isEmpty) {
+                  // Only makes sense to check that a binary expression is within bounds if the types of its
+                  // subexpressions can be combined
+
+                  // mergedType must exist because, otherwise typesAreMergeable.isEmpty would not hold
+                  val mergedType = typeMerge(l, r).get
+
+                  // The first two checks ensure that, if an operand is constant, then it must be assignable to the type
+                  // of the result. This makes the type system capable of rejecting expressions like `uint8(1) * (-1)`,
+                  // which are also rejected by the go compiler
+                  intExprWithinTypeBounds(n.left.asInstanceOf[PExpression], mergedType) ++
+                    intExprWithinTypeBounds(n.right.asInstanceOf[PExpression], mergedType) ++
+                    intExprWithinTypeBounds(n, mergedType)
+                } else noMessages
+              }
+              lIsInteger ++ rIsInteger ++ typesAreMergeable ++ exprWithinBounds
             }
+          case (_: PShiftLeft, l, r) =>
+            val integerOperands = assignableTo.errors(l, UNTYPED_INT_CONST)(n) ++ assignableTo.errors(r, UNTYPED_INT_CONST)(n)
+            if (integerOperands.isEmpty) {
+              intConstantEval(n.right.asInstanceOf[PExpression]) match {
+                case Some(v) =>
+                  // The Go compiler checks that the RHS of (<<) is non-negative and, at most, the size
+                  // of the type of the left operand (or 512 if there is an untyped const on the left)
+                  val lowerBound = error(n.right, s"constant ${n.right} overflows uint", v < 0)
```

Comment:
Is this indeed called `overflow`? A negative left shift sounds to me more like a right shift with non-negative value

Reply:
The Go compiler does indeed report this as an overflow. For example, in this program, 
```go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(1 << -1)
}
```
Go reports the following erros:
```
./prog.go:8:16: constant -1 overflows uint
./prog.go:8:16: invalid negative shift count: -1
```

The reason is that Go expects the right operand to be of uint type and considers that negative right operands overflow uint

Reply:
I see, thanks for the explanation!